app/models/spotlight/resource.rb

```diff
@@ -74,72 +35,141 @@ def needs_provider?
       type.blank?
     end
 
-    def save_and_commit
-      save.tap do
-        begin
-          blacklight_solr.commit
-        rescue => e
-          Rails.logger.warn "Unable to commit to solr: #{e}"
+    ##
+    # Persist the record to the database, and trigger a reindex to solr
+    #
+    # @param [Hash] All arguments will be passed through to ActiveRecord's #save method
+    def save_and_index(*args)
+      save(*args).tap { reindex }
+    end
+
+    concerning :GeneratingSolrDocuments do
+      ##
+      # @abstract
+      # Convert this resource into zero-to-many new solr documents. The data here
+      # should be merged into the resource-specific {#to_solr} data.
+      #
+      # @return [Hash] a single solr document hash
+      # @return [Enumerator<Hash>] multiple solr document hashes. This can be a
+      #   simple array, or an lazy enumerator
+      def to_solr
+        (exhibit_specific_solr_data || {}).merge(spotlight_resource_metadata_for_solr || {})
+      end
+
+      protected
+
+      ##
+      # @return an enumerator of all the indexable documents for this resource
+      def documents_to_index
+        data = to_solr
+        return [] if data.blank?
+        data &&= [data] if data.is_a? Hash
+
+        return to_enum(:documents_to_index) { data.size } unless block_given?
+
+        data.reject(&:blank?).each do |doc|
+          yield doc.reverse_merge(existing_solr_doc_hash(doc[unique_key]) || {})
         end
       end
-    end
 
-    protected
+      private
 
-    def blacklight_solr
-      @solr ||= RSolr.connect(connection_config)
-    end
+      ##
+      # Get any exhibit-specific metadata stored in e.g. sidecars, tags, etc
+      # This needs the generated solr document
+      def existing_solr_doc_hash(id)
+        document_model.new(unique_key => id).to_solr if document_model && id.present?
+      end
 
-    def connection_config
-      Blacklight.connection_config
-    end
+      def unique_key
+        if document_model
+          document_model.unique_key.to_sym
+        else
+          :id
+        end
+      end
 
-    def document_model
-      exhibit.blacklight_config.document_model if exhibit
-    end
+      def exhibit_specific_solr_data
+        exhibit.solr_data if exhibit
+      end
 
-    def batch_size
-      Spotlight::Engine.config.solr_batch_size
-    end
+      def spotlight_resource_metadata_for_solr
+        {
+          Spotlight::Engine.config.resource_global_id_field => (to_global_id.to_s if persisted?),
+          Spotlight::SolrDocument.resource_type_field => self.class.to_s.tableize
+        }
+      end
 
-    def exhibit_specific_solr_data
-      exhibit.solr_data if exhibit
+      def document_model
+        exhibit.blacklight_config.document_model if exhibit
+      end
```

Comment:
I'm a huge fan of  end # concern GeneratingSolrDocuments on the approp end statement when it's more than a screen.